cpp/include/cudf/types.hpp

```diff
@@ -68,6 +68,10 @@ template <typename T> class numeric_scalar;
 class string_scalar;
 template <typename T> class timestamp_scalar;
 
+namespace detail {
```

Comment:
Don't like this.  types.hpp needs to have forward declarations of everything.  The only thing I could think to put here was detail::scalar_device_view_base.  Perhaps that's fine... but it feels funny to me to use something from detail::

Reply:
I don't understand. What is causing the need to declare something from `detail::` in `types.hpp`?

Reply:
In type_dispatcher.hpp, line 120, we define type_to_scalar_type_impl::ScalarType to be cudf::scalar, the abstract type of all scalars.  We later specialize the template to set ScalarType to be the appropriate specialized type.

On line 121, I define the default value of type_to_scalar_type_impl::ScalarDeviceType to be cudf::detail::scalar_device_view_base with is the abstract type of all of the device views, and later specialize the template to set ScalarDeviceType to the appropriate value.

I'm not sure what the best solution is.  Seems to me we could:

1. Pull scalar_device_view_base out of the detail namespace,
2. Set the default value to something else (e.g. void *) since we should never instantiate type_to_scalar_type_impl from something that isn't specialized
3. Leave this as is (don't really like that option)

Reply:
See below. I think we should be able to do away with exposing `scalar_device_view_base` altogether.